### DISCUSSION TITLE
New cops and changes to existing cops for Ruby 3.0

### DISCUSSION THREAD

**Discussion by bbatsov** [rubocop/rubocop#9317] [Ideas]
Ruby 3.0 bundled quite a few changes, so I'm starting this topic as a place where we can discuss how to deal with them in RuboCop. I really liked this [detailed breakdown](https://rubyreferences.github.io/rubychanges/3.0.html) of the changes, as it gives us plenty of ideas for things to do (e.g. it seems that class vars are more reasonable now, which is something I had missed).

**Comment by marcandre** [comment]
The biggest change for RuboCop may be the availability of types, but I opened #9214 for this.

Class variables: Indeed, looks like class variables are a good spot to hold shared mutable data and all constants should be immutable. Even less exception to the new `#shareable_constant_values` pragma than I thought 👍.

Should we change `RuboCop::Cop::Registry.global`?

**Comment by bbatsov** [comment]
This might warrant a new cop as well - https://eregon.me/blog/2021/02/13/correct-delegation-in-ruby-2-27-3.html

**Reply by marcandre** [reply]
The post is good, but 99.9% of times it doesn't matter much; I don't see this bulletproof forwarding being implemented in the vast majority of cases.

**Reply by dvandersluis** [reply]
Yeah it probably only matters for gems because apps will be only targeting a single ruby version 99.9% of the time.

**Reply by marcandre** [reply]
Not only that, but most method forwarding happens for known destination methods. If a method does not accept keyword arguments, then forwarding with `(*args)` works fine. If it does, then `(*args, **options)` works fine. It's only in the case where you don't know the destination method you are forwarding to that is tricky.

**Reply by bbatsov** [reply]
Fair points. Less work for us then. :-)
